File: framework/src/play-cache/src/main/java/play/cache/AsyncCacheApi.java
/*
 * Copyright (C) 2009-2018 Lightbend Inc. <https://www.lightbend.com>
 */

package play.cache;

import java.util.concurrent.Callable;
import java.util.concurrent.CompletionStage;

import akka.Done;

/**
 * The Cache API.
 */
public interface AsyncCacheApi {

    /**
     * @return a synchronous version of this cache, which can be used to make synchronous calls.
     */
    default SyncCacheApi sync() {
        return new DefaultSyncCacheApi(this);
    }

    /**
     * Retrieves an object by key.
     *
     * @param <T> the type of the stored object
     * @param key the key to look up
     * @return a CompletionStage containing the value
     */
    <T> CompletionStage<T> get(String key);

    /**
     * Retrieve a value from the cache, or set it from a default Callable function.
     *
     * @param <T> the type of the value
     * @param key Item key.
     * @param block block returning value to set if key does not exist
     * @param expiration expiration period in seconds.
     * @return a CompletionStage containing the value
     */
    <T> CompletionStage<T> getOrElseUpdate(String key, Callable<CompletionStage<T>> block, int expiration);

    /**
     * Retrieve a value from the cache, or set it from a default Callable function.
     *
     * The value has no expiration.
     *
     * @param <T> the type of the value
     * @param key Item key.
     * @param block block returning value to set if key does not exist
     * @return a CompletionStage containing the value
     */
    <T> CompletionStage<T> getOrElseUpdate(String key, Callable<CompletionStage<T>> block);

    /**
     * Sets a value with expiration.
     *
     * @param key Item key.
     * @param value The value to set.
     * @param expiration expiration in seconds
     * @return a CompletionStage containing the value
     */
    CompletionStage<Done> set(String key, Object value, int expiration);

    /**
     * Sets a value without expiration.
     *
     * @param key Item key.
     * @param value The value to set.
     * @return a CompletionStage containing the value
     */
    CompletionStage<Done> set(String key, Object value);

    /**
     * Removes a value from the cache.
     *
     * @param key The key to remove the value for.
     * @return a CompletionStage containing the value
     */
    CompletionStage<Done> remove(String key);

    /**
     * Removes all values from the cache. This may be useful as an admin user operation if it is supported by your cache.
     *
     * @throws UnsupportedOperationException if this cache implementation does not support removing all values.
     * @return a CompletionStage containing either a Done when successful or an exception when unsuccessful.
     */
    CompletionStage<Done> removeAll();
}

File: framework/src/play-cache/src/main/scala/play/api/cache/AsyncCacheApi.scala
/*
 * Copyright (C) 2009-2018 Lightbend Inc. <https://www.lightbend.com>
 */

package play.api.cache

import akka.Done

import scala.concurrent.Future
import scala.concurrent.duration._
import scala.reflect.ClassTag

/**
 * The cache API
 */
trait AsyncCacheApi {

  /**
   * Get an instance of [[SyncCacheApi]] to make synchronous calls.
   */
  lazy val sync: SyncCacheApi = new DefaultSyncCacheApi(this)

  /**
   * Set a value into the cache.
   *
   * @param key Item key.
   * @param value Item value.
   * @param expiration Expiration time.
   */
  def set(key: String, value: Any, expiration: Duration = Duration.Inf): Future[Done]

  /**
   * Remove a value from the cache
   */
  def remove(key: String): Future[Done]

  /**
   * Retrieve a value from the cache, or set it from a default function.
   *
   * @param key Item key.
   * @param expiration expiration period in seconds.
   * @param orElse The default function to invoke if the value was not found in cache.
   */
  def getOrElseUpdate[A: ClassTag](key: String, expiration: Duration = Duration.Inf)(orElse: => Future[A]): Future[A]

  /**
   * Retrieve a value from the cache for the given type
   *
   * @param key Item key.
   * @return result as a future of Option[T]
   */
  def get[T: ClassTag](key: String): Future[Option[T]]

  /**
   * Removes all values from the cache. This may be useful as an admin user operation if it is supported by your cache.
   *
   * @throws UnsupportedOperationException if this cache implementation does not support removing all values.
   * @return a Future[Done], which is completed with either a Done or an exception if the clear did not work.
   */
  def removeAll(): Future[Done]
}
